File: DevExpress-Demo/CS/ChartsDemo.Wpf/DataModels/AgeStructure.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using System.Windows;
using System.Windows.Media;
using DevExpress.Mvvm.POCO;
using DevExpress.Xpf.Charts;

namespace ChartsDemo {
    public class AgeStructure {
        DataTable ageStructureTable;
        DataTable AgeStructureTable {
            get {
                if (this.ageStructureTable == null)
                    this.ageStructureTable = LoadPopulationAgeStructure();
                return this.ageStructureTable;
            }
        }

        DataTable LoadPopulationAgeStructure() {
            Uri uri = new Uri(@"pack://application:,,,/ChartsDemo;component/Data/Population.xml");
            Stream xmlStream = Application.GetResourceStream(uri).Stream;
            DataSet xmlDataSet = new DataSet();
            xmlDataSet.ReadXml(xmlStream);
            xmlStream.Close();
            return xmlDataSet.Tables["Population"];
        }

        public IList DataByAgeAndGender {
            get {
                return AgeStructureTable.AsEnumerable()
                    .Select(row => new {
                        GenderAge = new GenderAgeInfo(row.Field<string>("Gender"), row.Field<string>("Age")),
                        Country = row.Field<string>("Country"),
                        Population = row.Field<long>("Population")
                    }).ToList();
            }
        }
        public IList GetDataByMaleAge() {
            return AgeStructureTable.AsEnumerable()
                .Where(row => row.Field<string>("Gender") == "Male")
                .Select(row => new {
                    Age = row.Field<string>("Age"),
                    Country = row.Field<string>("Country"),
                    Population = row.Field<long>("Population")
                }).ToList();
        }
        public IList GetPopulationAgeStructure() {
            return AgeStructureTable.AsEnumerable()
                .Select(row => new AgePopulation(row.Field<string>("Country"), row.Field<string>("Age"), row.Field<string>("Gender"), row.Field<long>("Population"))).ToList();
        }

        public IList GetGenderAgeItemsWithPopulation() {
            Dictionary<string, List<PopulationItem>> genderAgePopulationItemPair = new Dictionary<string, List<PopulationItem>>();
            foreach (DataRow row in AgeStructureTable.AsEnumerable()) {
                string gender = row.Field<string>("Gender");
                long population = row.Field<long>("Population");
                string populationString = population.ToString("0,,.00");
                string country = row.Field<string>("Country");
                string age = row.Field<string>("Age");
                string key = gender + ": " + age;

                List<PopulationItem> populationItemsBuffer;
                if (genderAgePopulationItemPair.TryGetValue(key, out populationItemsBuffer)) {
                    populationItemsBuffer.Add(new PopulationItem(country, gender, population, populationString));
                }
                else {
                    populationItemsBuffer = new List<PopulationItem>();
                    populationItemsBuffer.Add(new PopulationItem(country, gender, population, populationString));
                    genderAgePopulationItemPair.Add(key, populationItemsBuffer);
                }
            }

            List<GenderAgeItem> genderAgeItems = new List<GenderAgeItem>();
            foreach (KeyValuePair<string, List<PopulationItem>> keyValue in genderAgePopulationItemPair) {
                genderAgeItems.Add(GenderAgeItem.Create(keyValue.Key, keyValue.Value));
            }
            return genderAgeItems;
        }
    }


    public struct GenderAgeInfo {
        readonly string gender;
        readonly string age;

        public string Gender { get { return this.gender; } }
        public string Age { get { return this.age; } }

        public GenderAgeInfo(string gender, string age) {
            this.gender = gender;
            this.age = age;
        }

        public override string ToString() {
            return this.gender + ": " + this.age;
        }
    }

    public class GenderAgeItem {
        public static GenderAgeItem Create(string name, IList<PopulationItem> items) {
            return ViewModelSource.Create(() => new GenderAgeItem(name, items));
        }
        public virtual string Name { get; protected set; }
        public virtual Color Color { get; set; }
        public virtual IList<PopulationItem> Items { get; protected set; }
        public virtual bool IsShowLabelsChecked { get; set; }
        public virtual bool IsShowToolTipsChecked { get; set; }
        public virtual Bar2DModel Model { get; set; }
        public virtual LegendViewModel Legend { get; set; }
        public virtual PaneViewModel Pane { get; set; }

        protected GenderAgeItem(string name, IList<PopulationItem> items) {
            Name = name;
            Items = items;
            IsShowToolTipsChecked = true;
        }
    }

    public struct PopulationItem {
        readonly string country;
        readonly string gender;
        readonly long population;
        readonly string populationString;

        public string Country { get { return country; } }
        public string Gender { get { return gender; } }
        public long Population { get { return population; } }
        public string PopulationString { get { return populationString; } }

        public PopulationItem(string country, string gender, long population, string populationString) {
            this.country = country;
            this.gender = gender;
            this.population = population;
            this.populationString = populationString;
        }
    }

    public class AgePopulation {
        readonly string name;
        readonly string age;
        readonly string sex;
        readonly double population;

        public string Name { get { return this.name; } }
        public string Age { get { return this.age; } }
        public string Sex { get { return this.sex; } }
        public string SexAgeKey { get { return this.sex.ToString() + ": " + this.age; } }
        public string CountryAgeKey { get { return this.name + ": " + this.age; } }
        public string CountrySexKey { get { return this.name + ": " + this.sex.ToString(); } }
        public double Population { get { return this.population; } }

        public AgePopulation(string name, string age, string gender, double population) {
            this.name = name;
            this.age = age;
            this.sex = gender;
            this.population = population;
        }
    }
}

File: DevExpress-Demo/VB/ChartsDemo.Wpf/DataModels/AgeStructure.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Data
Imports System.IO
Imports System.Linq
Imports System.Windows
Imports System.Windows.Media
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Xpf.Charts

Namespace ChartsDemo

    Public Class AgeStructure

        Private ageStructureTableField As DataTable

        Private ReadOnly Property AgeStructureTable As DataTable
            Get
                If ageStructureTableField Is Nothing Then ageStructureTableField = LoadPopulationAgeStructure()
                Return ageStructureTableField
            End Get
        End Property

        Private Function LoadPopulationAgeStructure() As DataTable
            Dim uri As Uri = New Uri("pack://application:,,,/ChartsDemo;component/Data/Population.xml")
            Dim xmlStream As Stream = Application.GetResourceStream(uri).Stream
            Dim xmlDataSet As DataSet = New DataSet()
            xmlDataSet.ReadXml(xmlStream)
            xmlStream.Close()
            Return xmlDataSet.Tables("Population")
        End Function

        Public ReadOnly Property DataByAgeAndGender As IList
            Get
                Return AgeStructureTable.AsEnumerable().[Select](Function(row) New With {.GenderAge = New GenderAgeInfo(row.Field(Of String)("Gender"), row.Field(Of String)("Age")), .Country = row.Field(Of String)("Country"), .Population = row.Field(Of Long)("Population")}).ToList()
            End Get
        End Property

        Public Function GetDataByMaleAge() As IList
            Return AgeStructureTable.AsEnumerable().Where(Function(row) Equals(row.Field(Of String)("Gender"), "Male")).[Select](Function(row) New With {.Age = row.Field(Of String)("Age"), .Country = row.Field(Of String)("Country"), .Population = row.Field(Of Long)("Population")}).ToList()
        End Function

        Public Function GetPopulationAgeStructure() As IList
            Return AgeStructureTable.AsEnumerable().[Select](Function(row) New AgePopulation(row.Field(Of String)("Country"), row.Field(Of String)("Age"), row.Field(Of String)("Gender"), row.Field(Of Long)("Population"))).ToList()
        End Function

        Public Function GetGenderAgeItemsWithPopulation() As IList
            Dim genderAgePopulationItemPair As Dictionary(Of String, List(Of PopulationItem)) = New Dictionary(Of String, List(Of PopulationItem))()
            For Each row As DataRow In AgeStructureTable.AsEnumerable()
                Dim gender As String = row.Field(Of String)("Gender")
                Dim population As Long = row.Field(Of Long)("Population")
                Dim populationString As String = population.ToString("0,,.00")
                Dim country As String = row.Field(Of String)("Country")
                Dim age As String = row.Field(Of String)("Age")
                Dim key As String = gender & ": " & age
                Dim populationItemsBuffer As List(Of PopulationItem)
                If genderAgePopulationItemPair.TryGetValue(key, populationItemsBuffer) Then
                    populationItemsBuffer.Add(New PopulationItem(country, gender, population, populationString))
                Else
                    populationItemsBuffer = New List(Of PopulationItem)()
                    populationItemsBuffer.Add(New PopulationItem(country, gender, population, populationString))
                    genderAgePopulationItemPair.Add(key, populationItemsBuffer)
                End If
            Next

            Dim genderAgeItems As List(Of GenderAgeItem) = New List(Of GenderAgeItem)()
            For Each keyValue As KeyValuePair(Of String, List(Of PopulationItem)) In genderAgePopulationItemPair
                genderAgeItems.Add(GenderAgeItem.Create(keyValue.Key, keyValue.Value))
            Next

            Return genderAgeItems
        End Function
    End Class

    Public Structure GenderAgeInfo

        Private ReadOnly genderField As String

        Private ReadOnly ageField As String

        Public ReadOnly Property Gender As String
            Get
                Return genderField
            End Get
        End Property

        Public ReadOnly Property Age As String
            Get
                Return ageField
            End Get
        End Property

        Public Sub New(ByVal gender As String, ByVal age As String)
            genderField = gender
            ageField = age
        End Sub

        Public Overrides Function ToString() As String
            Return genderField & ": " & ageField
        End Function
    End Structure

    Public Class GenderAgeItem

        Public Shared Function Create(ByVal name As String, ByVal items As IList(Of PopulationItem)) As GenderAgeItem
            Return ViewModelSource.Create(Function() New GenderAgeItem(name, items))
        End Function

        Public Overridable Property Name As String

        Public Overridable Property Color As Color

        Public Overridable Property Items As IList(Of PopulationItem)

        Public Overridable Property IsShowLabelsChecked As Boolean

        Public Overridable Property IsShowToolTipsChecked As Boolean

        Public Overridable Property Model As Bar2DModel

        Public Overridable Property Legend As LegendViewModel

        Public Overridable Property Pane As PaneViewModel

        Protected Sub New(ByVal name As String, ByVal items As IList(Of PopulationItem))
            Me.Name = name
            Me.Items = items
            IsShowToolTipsChecked = True
        End Sub
    End Class

    Public Structure PopulationItem

        Private ReadOnly countryField As String

        Private ReadOnly genderField As String

        Private ReadOnly populationField As Long

        Private ReadOnly populationStringField As String

        Public ReadOnly Property Country As String
            Get
                Return countryField
            End Get
        End Property

        Public ReadOnly Property Gender As String
            Get
                Return genderField
            End Get
        End Property

        Public ReadOnly Property Population As Long
            Get
                Return populationField
            End Get
        End Property

        Public ReadOnly Property PopulationString As String
            Get
                Return populationStringField
            End Get
        End Property

        Public Sub New(ByVal country As String, ByVal gender As String, ByVal population As Long, ByVal populationString As String)
            countryField = country
            genderField = gender
            populationField = population
            populationStringField = populationString
        End Sub
    End Structure

    Public Class AgePopulation

        Private ReadOnly nameField As String

        Private ReadOnly ageField As String

        Private ReadOnly sexField As String

        Private ReadOnly populationField As Double

        Public ReadOnly Property Name As String
            Get
                Return nameField
            End Get
        End Property

        Public ReadOnly Property Age As String
            Get
                Return ageField
            End Get
        End Property

        Public ReadOnly Property Sex As String
            Get
                Return sexField
            End Get
        End Property

        Public ReadOnly Property SexAgeKey As String
            Get
                Return sexField.ToString() & ": " & ageField
            End Get
        End Property

        Public ReadOnly Property CountryAgeKey As String
            Get
                Return nameField & ": " & ageField
            End Get
        End Property

        Public ReadOnly Property CountrySexKey As String
            Get
                Return nameField & ": " & sexField.ToString()
            End Get
        End Property

        Public ReadOnly Property Population As Double
            Get
                Return populationField
            End Get
        End Property

        Public Sub New(ByVal name As String, ByVal age As String, ByVal gender As String, ByVal population As Double)
            nameField = name
            ageField = age
            sexField = gender
            populationField = population
        End Sub
    End Class
End Namespace
